In [1]:
import pandas as pd
import numpy as np
import matplotlib.pylab as plt
import seaborn as sns
plt.style.use('ggplot')

In [2]:
data_path = "/home/tahas44/Desktop/NYC-Taxi/datas/yellow_tripdata_2015-01.parquet"
df = pd.read_parquet(data_path)
df.head()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,airport_fee
0,1,2015-01-01 00:11:33,2015-01-01 00:16:48,1,1.0,1,N,41,166,1,5.7,0.5,0.5,1.40,0.0,0.0,8.40,None,None
1,1,2015-01-01 00:18:24,2015-01-01 00:24:20,1,0.9,1,N,166,238,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30,None,None
2,1,2015-01-01 00:26:19,2015-01-01 00:41:06,1,3.5,1,N,238,162,1,13.2,0.5,0.5,2.90,0.0,0.0,17.40,None,None
3,1,2015-01-01 00:45:26,2015-01-01 00:53:20,1,2.1,1,N,162,263,1,8.2,0.5,0.5,2.37,0.0,0.0,11.87,None,None
4,1,2015-01-01 00:59:21,2015-01-01 01:05:24,1,1.0,1,N,236,141,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30,None,None


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12741035 entries, 0 to 12741034
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        int64         
 4   trip_distance          float64       
 5   RatecodeID             int64         
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   object        
 18  airport_fee         

### Step 1: Data Understanding
- Dataframe **shape**
- **head** and **tail**
- **dtypes**
- **describe**

In [4]:
df.shape

(12741035, 19)

In [5]:
df.columns

Index(['VendorID', 'tpep_pickup_datetime', 'tpep_dropoff_datetime',
       'passenger_count', 'trip_distance', 'RatecodeID', 'store_and_fwd_flag',
       'PULocationID', 'DOLocationID', 'payment_type', 'fare_amount', 'extra',
       'mta_tax', 'tip_amount', 'tolls_amount', 'improvement_surcharge',
       'total_amount', 'congestion_surcharge', 'airport_fee'],
      dtype='object')

In [6]:
df.dtypes

VendorID                          int64
tpep_pickup_datetime     datetime64[us]
tpep_dropoff_datetime    datetime64[us]
passenger_count                   int64
trip_distance                   float64
RatecodeID                        int64
store_and_fwd_flag               object
PULocationID                      int64
DOLocationID                      int64
payment_type                      int64
fare_amount                     float64
extra                           float64
mta_tax                         float64
tip_amount                      float64
tolls_amount                    float64
improvement_surcharge           float64
total_amount                    float64
congestion_surcharge             object
airport_fee                      object
dtype: object

In [7]:
df.describe()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
count,1.274104e+07,12741035,12741035,1.274104e+07,1.274104e+07,1.274104e+07,1.274104e+07,1.274104e+07,1.274104e+07,1.274104e+07,1.274104e+07,1.274104e+07,1.274104e+07,1.274104e+07,1.274103e+07,1.274104e+07
mean,1.521504e+00,2015-01-16 14:42:27.972079,2015-01-16 14:56:38.686013,1.681725e+00,1.312720e+01,1.036635e+00,1.632477e+02,1.610789e+02,1.386464e+00,1.190480e+01,3.082480e-01,4.978640e-01,1.854764e+00,2.433988e-01,2.831486e-01,1.510835e+01
min,1.000000e+00,2015-01-01 00:00:00,2015-01-01 00:00:00,0.000000e+00,0.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,-4.500000e+02,-7.900000e+01,-5.000000e-01,-9.242000e+01,-2.600000e+01,0.000000e+00,-4.503000e+02
25%,1.000000e+00,2015-01-09 09:26:01.500000,2015-01-09 09:43:41.500000,1.000000e+00,1.000000e+00,1.000000e+00,1.130000e+02,1.070000e+02,1.000000e+00,6.500000e+00,0.000000e+00,5.000000e-01,0.000000e+00,0.000000e+00,3.000000e-01,8.160000e+00
50%,2.000000e+00,2015-01-16 11:43:24,2015-01-16 11:57:36,1.000000e+00,1.680000e+00,1.000000e+00,1.620000e+02,1.620000e+02,1.000000e+00,9.000000e+00,0.000000e+00,5.000000e-01,1.000000e+00,0.000000e+00,3.000000e-01,1.116000e+01
75%,2.000000e+00,2015-01-23 19:15:56,2015-01-23 19:29:29,2.000000e+00,3.000000e+00,1.000000e+00,2.330000e+02,2.330000e+02,2.000000e+00,1.350000e+01,5.000000e-01,5.000000e-01,2.060000e+00,0.000000e+00,3.000000e-01,1.630000e+01
max,2.000000e+00,2015-01-31 23:59:59,2016-02-02 16:30:52,9.000000e+00,1.542000e+07,9.900000e+01,2.650000e+02,2.650000e+02,5.000000e+00,4.008000e+03,6.480200e+02,5.000000e-01,3.950589e+06,1.450090e+03,3.000000e-01,3.950612e+06
std,4.995374e-01,NaN,NaN,1.338111e+00,9.821939e+03,6.677706e-01,6.726112e+01,7.079588e+01,4.987584e-01,1.020520e+01,5.053308e-01,3.487909e-02,1.106777e+03,1.474659e+00,6.907570e-02,1.106847e+03


### Step 2: Data Preperation
- Dropping irrelevant columns and rows
- Identifying duplicated columns
- Renaming columns
- Feature Creation

In [8]:
df.head()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,airport_fee
0,1,2015-01-01 00:11:33,2015-01-01 00:16:48,1,1.0,1,N,41,166,1,5.7,0.5,0.5,1.40,0.0,0.0,8.40,None,None
1,1,2015-01-01 00:18:24,2015-01-01 00:24:20,1,0.9,1,N,166,238,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30,None,None
2,1,2015-01-01 00:26:19,2015-01-01 00:41:06,1,3.5,1,N,238,162,1,13.2,0.5,0.5,2.90,0.0,0.0,17.40,None,None
3,1,2015-01-01 00:45:26,2015-01-01 00:53:20,1,2.1,1,N,162,263,1,8.2,0.5,0.5,2.37,0.0,0.0,11.87,None,None
4,1,2015-01-01 00:59:21,2015-01-01 01:05:24,1,1.0,1,N,236,141,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30,None,None


In [9]:
df.drop(["VendorID"], axis=1)   # görmek istemediğin sütunu kaldır

,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,airport_fee
0,2015-01-01 00:11:33,2015-01-01 00:16:48,1,1.0,1,N,41,166,1,5.7,0.5,0.5,1.40,0.0,0.0,8.40,None,None
1,2015-01-01 00:18:24,2015-01-01 00:24:20,1,0.9,1,N,166,238,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30,None,None
2,2015-01-01 00:26:19,2015-01-01 00:41:06,1,3.5,1,N,238,162,1,13.2,0.5,0.5,2.90,0.0,0.0,17.40,None,None
3,2015-01-01 00:45:26,2015-01-01 00:53:20,1,2.1,1,N,162,263,1,8.2,0.5,0.5,2.37,0.0,0.0,11.87,None,None
4,2015-01-01 00:59:21,2015-01-01 01:05:24,1,1.0,1,N,236,141,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30,None,None
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12741030,2015-01-31 23:21:42,2015-01-31 23:31:00,1,1.6,1,N,90,249,1,8.0,0.5,0.5,2.32,0.0,0.3,11.62,None,None
12741031,2015-01-31 23:42:43,2015-01-31 23:49:32,1,0.6,1,N,90,68,1,6.0,0.5,0.5,1.46,0.0,0.3,8.76,None,None
12741032,2015-01-31 23:55:16,2015-02-01 00:16:45,1,3.0,1,N,68,148,1,15.0,0.5,0.5,4.07,0.0,0.3,20.37,None,None
12741033,2015-01-31 23:20:53,2015-02-01 00:07:35,1,6.9,1,N,189,237,1,32.5,0.5,0.5,6.00,0.0,0.3,39.80,None,None


In [10]:
df[df["airport_fee"].notna()] 

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,airport_fee


görünüşe göre airport_fee dediğimiz havalimanı ücreti sütununda dolu bir sütun yok.

In [11]:
df.drop(columns=["airport_fee"], inplace=True)  # "airport_fee" hücresinden kurtuldum

In [12]:
df.head()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge
0,1,2015-01-01 00:11:33,2015-01-01 00:16:48,1,1.0,1,N,41,166,1,5.7,0.5,0.5,1.40,0.0,0.0,8.40,None
1,1,2015-01-01 00:18:24,2015-01-01 00:24:20,1,0.9,1,N,166,238,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30,None
2,1,2015-01-01 00:26:19,2015-01-01 00:41:06,1,3.5,1,N,238,162,1,13.2,0.5,0.5,2.90,0.0,0.0,17.40,None
3,1,2015-01-01 00:45:26,2015-01-01 00:53:20,1,2.1,1,N,162,263,1,8.2,0.5,0.5,2.37,0.0,0.0,11.87,None
4,1,2015-01-01 00:59:21,2015-01-01 01:05:24,1,1.0,1,N,236,141,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30,None


In [13]:
df.columns

Index(['VendorID', 'tpep_pickup_datetime', 'tpep_dropoff_datetime',
       'passenger_count', 'trip_distance', 'RatecodeID', 'store_and_fwd_flag',
       'PULocationID', 'DOLocationID', 'payment_type', 'fare_amount', 'extra',
       'mta_tax', 'tip_amount', 'tolls_amount', 'improvement_surcharge',
       'total_amount', 'congestion_surcharge'],
      dtype='object')

In [14]:
df[df["congestion_surcharge"].notna()]  # aynı airport_fee de karşılaştığım gibi burada da her şey None

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge


In [15]:
df.drop(columns=["congestion_surcharge"], inplace=True)

In [16]:
df

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
0,1,2015-01-01 00:11:33,2015-01-01 00:16:48,1,1.0,1,N,41,166,1,5.7,0.5,0.5,1.40,0.0,0.0,8.40
1,1,2015-01-01 00:18:24,2015-01-01 00:24:20,1,0.9,1,N,166,238,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30
2,1,2015-01-01 00:26:19,2015-01-01 00:41:06,1,3.5,1,N,238,162,1,13.2,0.5,0.5,2.90,0.0,0.0,17.40
3,1,2015-01-01 00:45:26,2015-01-01 00:53:20,1,2.1,1,N,162,263,1,8.2,0.5,0.5,2.37,0.0,0.0,11.87
4,1,2015-01-01 00:59:21,2015-01-01 01:05:24,1,1.0,1,N,236,141,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12741030,1,2015-01-31 23:21:42,2015-01-31 23:31:00,1,1.6,1,N,90,249,1,8.0,0.5,0.5,2.32,0.0,0.3,11.62
12741031,1,2015-01-31 23:42:43,2015-01-31 23:49:32,1,0.6,1,N,90,68,1,6.0,0.5,0.5,1.46,0.0,0.3,8.76
12741032,1,2015-01-31 23:55:16,2015-02-01 00:16:45,1,3.0,1,N,68,148,1,15.0,0.5,0.5,4.07,0.0,0.3,20.37
12741033,1,2015-01-31 23:20:53,2015-02-01 00:07:35,1,6.9,1,N,189,237,1,32.5,0.5,0.5,6.00,0.0,0.3,39.80


In [17]:
df[df["tolls_amount"] != 0.0]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
17,1,2015-01-01 00:35:34,2015-01-01 00:58:57,1,4.40,1,N,158,145,1,18.2,0.5,0.5,3.00,5.33,0.0,27.83
239,2,2015-01-01 00:58:11,2015-01-01 01:27:27,1,11.88,1,N,163,200,2,36.5,0.5,0.5,0.00,2.70,0.3,40.50
272,2,2015-01-01 00:36:57,2015-01-01 00:54:16,1,6.95,1,N,79,157,2,21.0,0.5,0.5,0.00,5.33,0.3,27.63
334,2,2015-01-01 00:52:37,2015-01-01 01:16:59,1,5.17,1,N,90,112,1,20.0,0.5,0.5,7.75,5.33,0.3,34.38
349,1,2015-01-01 00:25:32,2015-01-01 00:47:29,2,10.90,1,N,236,40,2,31.5,0.5,0.5,0.00,5.33,0.0,38.13
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12740924,1,2015-01-31 23:05:21,2015-01-31 23:30:27,1,9.70,1,N,138,238,1,30.0,0.5,0.5,5.80,5.33,0.3,42.43
12740942,1,2015-01-31 23:31:16,2015-02-01 00:24:51,1,15.20,1,N,48,241,2,48.5,0.5,0.5,0.00,2.44,0.3,52.24
12741009,1,2015-01-31 23:19:19,2015-02-01 00:15:08,3,13.00,1,N,163,220,1,49.5,0.5,0.5,15.00,4.88,0.3,70.68
12741015,2,2015-01-31 23:27:10,2015-02-01 00:07:33,1,12.72,1,N,114,55,1,40.5,0.5,0.5,5.00,5.33,0.3,52.13


In [18]:
df.isna().sum() # ne kadar NA değerlerim var?

VendorID                 0
tpep_pickup_datetime     0
tpep_dropoff_datetime    0
passenger_count          0
trip_distance            0
RatecodeID               0
store_and_fwd_flag       0
PULocationID             0
DOLocationID             0
payment_type             0
fare_amount              0
extra                    0
mta_tax                  0
tip_amount               0
tolls_amount             0
improvement_surcharge    3
total_amount             0
dtype: int64

eksik verim yok ama mantıksal açıdan yanlış veriler (price < 0) olabilir

In [19]:
df

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
0,1,2015-01-01 00:11:33,2015-01-01 00:16:48,1,1.0,1,N,41,166,1,5.7,0.5,0.5,1.40,0.0,0.0,8.40
1,1,2015-01-01 00:18:24,2015-01-01 00:24:20,1,0.9,1,N,166,238,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30
2,1,2015-01-01 00:26:19,2015-01-01 00:41:06,1,3.5,1,N,238,162,1,13.2,0.5,0.5,2.90,0.0,0.0,17.40
3,1,2015-01-01 00:45:26,2015-01-01 00:53:20,1,2.1,1,N,162,263,1,8.2,0.5,0.5,2.37,0.0,0.0,11.87
4,1,2015-01-01 00:59:21,2015-01-01 01:05:24,1,1.0,1,N,236,141,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12741030,1,2015-01-31 23:21:42,2015-01-31 23:31:00,1,1.6,1,N,90,249,1,8.0,0.5,0.5,2.32,0.0,0.3,11.62
12741031,1,2015-01-31 23:42:43,2015-01-31 23:49:32,1,0.6,1,N,90,68,1,6.0,0.5,0.5,1.46,0.0,0.3,8.76
12741032,1,2015-01-31 23:55:16,2015-02-01 00:16:45,1,3.0,1,N,68,148,1,15.0,0.5,0.5,4.07,0.0,0.3,20.37
12741033,1,2015-01-31 23:20:53,2015-02-01 00:07:35,1,6.9,1,N,189,237,1,32.5,0.5,0.5,6.00,0.0,0.3,39.80


In [20]:
df[df.duplicated()]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
495485,1,2015-01-02 12:45:27,2015-01-02 12:57:28,1,2.30,1,N,186,142,2,10.5,0.0,0.5,0.0,0.0,0.0,11.3
539505,2,2015-01-02 15:14:46,2015-01-02 15:16:17,1,0.00,1,N,264,264,2,2.1,0.0,0.5,0.0,0.0,0.3,2.9
542584,2,2015-01-02 15:14:46,2015-01-02 15:16:17,1,0.00,1,N,264,264,2,2.1,0.0,0.5,0.0,0.0,0.3,2.9
543221,2,2015-01-02 15:14:46,2015-01-02 15:16:17,1,0.00,1,N,264,264,2,2.1,0.0,0.5,0.0,0.0,0.3,2.9
545806,2,2015-01-02 15:14:46,2015-01-02 15:16:17,1,0.00,1,N,264,264,2,2.1,0.0,0.5,0.0,0.0,0.3,2.9
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5572083,2,2015-01-14 16:13:52,2015-01-14 16:14:01,1,0.00,1,N,264,264,2,2.0,0.0,0.5,0.0,0.0,0.3,2.8
5573035,2,2015-01-14 16:13:52,2015-01-14 16:14:01,1,0.00,1,N,264,264,2,2.0,0.0,0.5,0.0,0.0,0.3,2.8
5578913,2,2015-01-14 16:13:52,2015-01-14 16:14:01,1,0.00,1,N,264,264,2,2.0,0.0,0.5,0.0,0.0,0.3,2.8
5579596,2,2015-01-14 16:13:52,2015-01-14 16:14:01,1,0.00,1,N,264,264,2,2.0,0.0,0.5,0.0,0.0,0.3,2.8


In [21]:
# 229 tane duplicate veri ile karşılaştıktan sonra bunlardan kurtulmam gerek
df.drop_duplicates(inplace=True)

In [22]:
df.duplicated().sum()

np.int64(0)

In [23]:
df.shape

(12740806, 17)

### Data Cleaning for Logic

In [24]:
# df.describe()

In [25]:
df["trip_distance"].max()   # max trip distance değeri gereğinden fazla

np.float64(15420004.5)

In [26]:
df["RatecodeID"].max()  # RatecodeID en fazla 6 olabilir

np.int64(99)

In [27]:
df["fare_amount"].max()

np.float64(4008.0)

In [28]:
df["extra"].max()

np.float64(648.02)

In [29]:
df["mta_tax"].max()

np.float64(0.5)

In [30]:
df["tip_amount"].max()

np.float64(3950588.8)

In [31]:
df["tolls_amount"].max()

np.float64(1450.09)

Data Cleaning için 6 durum tespit ettim:

1. "fare_amount", "extra", "mta_tax", "tip_amount", "tolls_amount" bazı yerlerde negatif; bazı yerlerde ise çok büyük çıkmış. Bu değerler 
optimal bir değer ve 0 arasında olmalı!

### 1.1 fare_amount

In [32]:
# plt.figure(figsize=(12,6))
# sns.ecdfplot(data=df, x="fare_amount", linewidth=3, color="darkred")    # Empirical Cumulative Distribution Function
# plt.axvline(x=100, color="blue", linestyle="--", linewidth=2, label="The limit is 100") # limit 100
# plt.title("the distribution of the fare amout using a limit")
# plt.xlabel("The price", fontsize=14)
# plt.ylabel("The percantage of the data", fontsize=14)
# plt.legend()
# plt.show()

In [33]:
# df_sofar = df[df["fare_amount"] > 100]
# far_quantatity = len(df_sofar)
# original_size = len(df)
# perctange_far_amount = 100 * (far_quantatity / original_size)
# print(f"100 den fazla fare amount içeren satirlarin yüzdesi: {perctange_far_amount}")

Sonuç: yüzdelik olarak baktığımda fare_amount "100" den fazla olan değerler çok az. Şutlanabilir.

### 1.2 extra

In [34]:
df.describe()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
count,1.274081e+07,12740806,12740806,1.274081e+07,1.274081e+07,1.274081e+07,1.274081e+07,1.274081e+07,1.274081e+07,1.274081e+07,1.274081e+07,1.274081e+07,1.274081e+07,1.274081e+07,1.274080e+07,1.274081e+07
mean,1.521495e+00,2015-01-16 14:42:33.074415,2015-01-16 14:56:43.802299,1.681736e+00,1.312744e+01,1.036635e+00,1.632459e+02,1.610771e+02,1.386453e+00,1.190497e+01,3.082536e-01,4.978640e-01,1.854797e+00,2.434032e-01,2.831483e-01,1.510857e+01
min,1.000000e+00,2015-01-01 00:00:00,2015-01-01 00:00:00,0.000000e+00,0.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,-4.500000e+02,-7.900000e+01,-5.000000e-01,-9.242000e+01,-2.600000e+01,0.000000e+00,-4.503000e+02
25%,1.000000e+00,2015-01-09 09:25:53,2015-01-09 09:43:34,1.000000e+00,1.000000e+00,1.000000e+00,1.130000e+02,1.070000e+02,1.000000e+00,6.500000e+00,0.000000e+00,5.000000e-01,0.000000e+00,0.000000e+00,3.000000e-01,8.160000e+00
50%,2.000000e+00,2015-01-16 11:43:43,2015-01-16 11:57:52,1.000000e+00,1.680000e+00,1.000000e+00,1.620000e+02,1.620000e+02,1.000000e+00,9.000000e+00,0.000000e+00,5.000000e-01,1.000000e+00,0.000000e+00,3.000000e-01,1.116000e+01
75%,2.000000e+00,2015-01-23 19:16:03,2015-01-23 19:29:33,2.000000e+00,3.000000e+00,1.000000e+00,2.330000e+02,2.330000e+02,2.000000e+00,1.350000e+01,5.000000e-01,5.000000e-01,2.060000e+00,0.000000e+00,3.000000e-01,1.630000e+01
max,2.000000e+00,2015-01-31 23:59:59,2016-02-02 16:30:52,9.000000e+00,1.542000e+07,9.900000e+01,2.650000e+02,2.650000e+02,5.000000e+00,4.008000e+03,6.480200e+02,5.000000e-01,3.950589e+06,1.450090e+03,3.000000e-01,3.950612e+06
std,4.995378e-01,NaN,NaN,1.338120e+00,9.822027e+03,6.677766e-01,6.726038e+01,7.079517e+01,4.987561e-01,1.020520e+01,5.053337e-01,3.487941e-02,1.106787e+03,1.474672e+00,6.907624e-02,1.106857e+03


In [35]:
df[df["extra"] < 0]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
1155,2,2015-01-01 00:37:47,2015-01-01 00:40:07,2,0.07,1,N,170,170,3,-3.0,-0.5,-0.5,0.0,0.0,0.3,-4.3
7470,2,2015-01-01 00:02:30,2015-01-01 00:02:41,1,0.00,1,N,132,264,3,-2.5,-0.5,-0.5,0.0,0.0,0.3,-3.8
8577,2,2015-01-01 00:48:01,2015-01-01 00:48:22,1,0.03,1,N,24,24,2,-2.5,-0.5,-0.5,0.0,0.0,0.3,-3.8
8704,2,2015-01-01 00:49:05,2015-01-01 00:49:50,2,0.01,1,N,136,136,4,-2.5,-0.5,-0.5,0.0,0.0,0.3,-3.8
12460,2,2015-01-01 00:55:56,2015-01-01 01:03:51,4,0.44,1,N,75,75,4,-6.5,-0.5,-0.5,0.0,0.0,0.3,-7.8
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12691046,2,2015-01-31 22:39:53,2015-01-31 22:41:17,1,0.10,1,N,236,43,4,-3.0,-0.5,-0.5,0.0,0.0,0.3,-4.3
12697269,2,2015-01-31 22:37:09,2015-01-31 22:39:07,2,0.31,1,N,247,247,3,-3.5,-0.5,-0.5,0.0,0.0,0.3,-4.8
12697725,2,2015-01-31 22:46:06,2015-01-31 22:50:09,1,0.77,1,N,238,143,4,-5.0,-0.5,-0.5,0.0,0.0,0.3,-6.3
12698517,2,2015-01-31 22:45:54,2015-01-31 22:52:22,1,0.00,1,N,90,90,2,-5.5,-0.5,-0.5,0.0,0.0,0.3,-6.8


In [36]:
df["extra"].mean()

np.float64(0.30825357909067896)

In [37]:
# df_extra = df[df["extra"] > 1]
# extra_quantatity = len(df_extra)
# perctange_extra = 100 * (extra_quantatity / original_size)
# print(f"1 dolardan fazla olan extra yüzdesi: {perctange_extra}")

In [38]:
# plt.figure(figsize=(12,6))
# sns.ecdfplot(data=df, x="extra", linewidth=3, color="darkred")    # Empirical Cumulative Distribution Function
# plt.axvline(x=1, color="blue", linestyle="--", linewidth=2, label="The limit is 1") # limit 1
# plt.title("the distribution of the fare amout using a limit")
# plt.xlabel("The price", fontsize=14)
# plt.ylabel("The percantage of the data", fontsize=14)
# plt.legend()
# plt.show()

Sonuç: **0 <= extra <= 1**

### 1.3 mta_tax

In [39]:
# df.describe()

In [40]:
df[df["mta_tax"] < 0]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
1155,2,2015-01-01 00:37:47,2015-01-01 00:40:07,2,0.07,1,N,170,170,3,-3.0,-0.5,-0.5,0.0,0.0,0.3,-4.3
7470,2,2015-01-01 00:02:30,2015-01-01 00:02:41,1,0.00,1,N,132,264,3,-2.5,-0.5,-0.5,0.0,0.0,0.3,-3.8
8577,2,2015-01-01 00:48:01,2015-01-01 00:48:22,1,0.03,1,N,24,24,2,-2.5,-0.5,-0.5,0.0,0.0,0.3,-3.8
8704,2,2015-01-01 00:49:05,2015-01-01 00:49:50,2,0.01,1,N,136,136,4,-2.5,-0.5,-0.5,0.0,0.0,0.3,-3.8
12460,2,2015-01-01 00:55:56,2015-01-01 01:03:51,4,0.44,1,N,75,75,4,-6.5,-0.5,-0.5,0.0,0.0,0.3,-7.8
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12697725,2,2015-01-31 22:46:06,2015-01-31 22:50:09,1,0.77,1,N,238,143,4,-5.0,-0.5,-0.5,0.0,0.0,0.3,-6.3
12698517,2,2015-01-31 22:45:54,2015-01-31 22:52:22,1,0.00,1,N,90,90,2,-5.5,-0.5,-0.5,0.0,0.0,0.3,-6.8
12705028,2,2015-01-31 22:32:26,2015-01-31 22:32:31,1,0.00,2,N,264,264,4,-52.0,0.0,-0.5,0.0,0.0,0.3,-52.8
12705306,2,2015-01-31 22:08:30,2015-01-31 22:22:27,1,0.81,1,N,79,79,4,-9.5,-0.5,-0.5,0.0,0.0,0.3,-10.8


In [41]:
# plt.figure(figsize=(12,6))
# sns.ecdfplot(data=df, x="mta_tax", linewidth=3, color="darkred")    # Empirical Cumulative Distribution Function
# plt.axvline(x=0.5, color="blue", linestyle="--", linewidth=2, label="The limit is 0.5") # limit 0.5
# plt.title("the distribution of the fare amout using a limit")
# plt.xlabel("The price", fontsize=14)
# plt.ylabel("The percantage of the data", fontsize=14)
# plt.legend()
# plt.show()

In [42]:
# df_mta_tax = df[df["mta_tax"] > 0.5]
# mta_tax_quantity = len(df_mta_tax)
# perctange_mta_tax = 100 * (mta_tax_quantity / original_size)
# print(f"0.5 den fazla olan vergilerin yüzdesi: {perctange_mta_tax}")

In [43]:
# df_mta_tax = df[df["mta_tax"] > 0.4]
# mta_tax_quantity = len(df_mta_tax)
# perctange_mta_tax = 100 * (mta_tax_quantity / original_size)
# print(f"0.4 den fazla olan vergilerin yüzdesi: {perctange_mta_tax}")

In [44]:
# df_mta_tax = df[(df["mta_tax"] < 0.4) & (df["mta_tax"] > 0)]
# mta_tax_quantity = len(df_mta_tax)
# perctange_mta_tax = 100 * (mta_tax_quantity / original_size)
# print(f"0 < x < 0.4 olan vergilerin yüzdesi: {perctange_mta_tax}")

In [45]:
df["mta_tax"]

0           0.5
1           0.5
2           0.5
3           0.5
4           0.5
           ... 
12741030    0.5
12741031    0.5
12741032    0.5
12741033    0.5
12741034    0.5
Name: mta_tax, Length: 12740806, dtype: float64

Sonuç: Negatif değerleri saymazsak "mta_tax" için değerler 0.4-0.5 arasında, bu metrik için fazladan üst sınır belirlemeye gerek yok. Negatif değerleri temizle üst sınırı olduğu gibi bırak

### 1.4 tip amount

In [46]:
# df.describe()

In [47]:
# df_nonnegative_tip_amount = df[df["tip_amount"] >= 0]
# df_nonnegative_tip_amount["tip_amount"].mean()
df["tip_amount"].mean()

np.float64(1.8547970622894645)

In [48]:
# plt.figure(figsize=(12,6))
# sns.ecdfplot(data=df, x="tip_amount", linewidth=3, color="darkred")    # Empirical Cumulative Distribution Function
# plt.axvline(x=10, color="blue", linestyle="--", linewidth=2, label="The limit is 10") # limit 10
# plt.title("the distribution of the tip amout using a limit")
# plt.xlabel("The price", fontsize=14)
# plt.ylabel("The percantage of the data", fontsize=14)
# plt.legend()
# plt.show()

In [49]:
# df_negative = df[df["tip_amount"] < 0]
# negative_quantaty = len(df_negative)
# negative_quantaty

In [50]:
# df_nonnegative_tip_amount = df[df["tip_amount"] > 1]
# original_size = len(df)
# original_size -= negative_quantaty
# percantage = 100 * (len(df_nonnegative_tip_amount) / original_size)
# print(f"1 den fazla olan bahşişler: {percantage}")

In [51]:
# df_nonnegative_tip_amount = df[(df["tip_amount"] < 1) & (df["tip_amount"] > 0)]
# original_size = len(df)
# original_size -= negative_quantaty
# percantage = 100 * (len(df_nonnegative_tip_amount) / original_size)
# print(f"0 < x < 1 olan bahşişler: {percantage}")

In [52]:
df["tip_amount"].max()

np.float64(3950588.8)

In [53]:
# df_nonnegative_tip_amount = df[df["tip_amount"] > 20]
# original_size = len(df)
# original_size -= negative_quantaty
# percantage = 100 * (len(df_nonnegative_tip_amount) / original_size)
# print(f"20 den fazla olan bahşişler: {percantage}")

In [54]:
df["tip_amount"].quantile(0.99) # quantile ile orantısal olarak istediğim üst sınırı bulabilirim

np.float64(11.3)

In [55]:
df["tip_amount"].quantile(0.999)

np.float64(17.44)

In [56]:
# df_tip_amount = df[df["tip_amount"] < 17.44]
# tip_amount_quanitity = len(df_tip_amount)
# percantage = 100 * (tip_amount_quanitity / len(df))
# print(f"17.44 den az olanlain yüzdesi: {percantage}")

Sonuç: 0 < x < 18 tip amount için uygulanabilir bir aralık

### 1.5 Tolls amount

In [57]:
df

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
0,1,2015-01-01 00:11:33,2015-01-01 00:16:48,1,1.0,1,N,41,166,1,5.7,0.5,0.5,1.40,0.0,0.0,8.40
1,1,2015-01-01 00:18:24,2015-01-01 00:24:20,1,0.9,1,N,166,238,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30
2,1,2015-01-01 00:26:19,2015-01-01 00:41:06,1,3.5,1,N,238,162,1,13.2,0.5,0.5,2.90,0.0,0.0,17.40
3,1,2015-01-01 00:45:26,2015-01-01 00:53:20,1,2.1,1,N,162,263,1,8.2,0.5,0.5,2.37,0.0,0.0,11.87
4,1,2015-01-01 00:59:21,2015-01-01 01:05:24,1,1.0,1,N,236,141,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12741030,1,2015-01-31 23:21:42,2015-01-31 23:31:00,1,1.6,1,N,90,249,1,8.0,0.5,0.5,2.32,0.0,0.3,11.62
12741031,1,2015-01-31 23:42:43,2015-01-31 23:49:32,1,0.6,1,N,90,68,1,6.0,0.5,0.5,1.46,0.0,0.3,8.76
12741032,1,2015-01-31 23:55:16,2015-02-01 00:16:45,1,3.0,1,N,68,148,1,15.0,0.5,0.5,4.07,0.0,0.3,20.37
12741033,1,2015-01-31 23:20:53,2015-02-01 00:07:35,1,6.9,1,N,189,237,1,32.5,0.5,0.5,6.00,0.0,0.3,39.80


In [58]:
df.describe()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
count,1.274081e+07,12740806,12740806,1.274081e+07,1.274081e+07,1.274081e+07,1.274081e+07,1.274081e+07,1.274081e+07,1.274081e+07,1.274081e+07,1.274081e+07,1.274081e+07,1.274081e+07,1.274080e+07,1.274081e+07
mean,1.521495e+00,2015-01-16 14:42:33.074415,2015-01-16 14:56:43.802299,1.681736e+00,1.312744e+01,1.036635e+00,1.632459e+02,1.610771e+02,1.386453e+00,1.190497e+01,3.082536e-01,4.978640e-01,1.854797e+00,2.434032e-01,2.831483e-01,1.510857e+01
min,1.000000e+00,2015-01-01 00:00:00,2015-01-01 00:00:00,0.000000e+00,0.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,-4.500000e+02,-7.900000e+01,-5.000000e-01,-9.242000e+01,-2.600000e+01,0.000000e+00,-4.503000e+02
25%,1.000000e+00,2015-01-09 09:25:53,2015-01-09 09:43:34,1.000000e+00,1.000000e+00,1.000000e+00,1.130000e+02,1.070000e+02,1.000000e+00,6.500000e+00,0.000000e+00,5.000000e-01,0.000000e+00,0.000000e+00,3.000000e-01,8.160000e+00
50%,2.000000e+00,2015-01-16 11:43:43,2015-01-16 11:57:52,1.000000e+00,1.680000e+00,1.000000e+00,1.620000e+02,1.620000e+02,1.000000e+00,9.000000e+00,0.000000e+00,5.000000e-01,1.000000e+00,0.000000e+00,3.000000e-01,1.116000e+01
75%,2.000000e+00,2015-01-23 19:16:03,2015-01-23 19:29:33,2.000000e+00,3.000000e+00,1.000000e+00,2.330000e+02,2.330000e+02,2.000000e+00,1.350000e+01,5.000000e-01,5.000000e-01,2.060000e+00,0.000000e+00,3.000000e-01,1.630000e+01
max,2.000000e+00,2015-01-31 23:59:59,2016-02-02 16:30:52,9.000000e+00,1.542000e+07,9.900000e+01,2.650000e+02,2.650000e+02,5.000000e+00,4.008000e+03,6.480200e+02,5.000000e-01,3.950589e+06,1.450090e+03,3.000000e-01,3.950612e+06
std,4.995378e-01,NaN,NaN,1.338120e+00,9.822027e+03,6.677766e-01,6.726038e+01,7.079517e+01,4.987561e-01,1.020520e+01,5.053337e-01,3.487941e-02,1.106787e+03,1.474672e+00,6.907624e-02,1.106857e+03


In [59]:
df[df["tolls_amount"] < 0]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
652852,2,2015-01-02 20:03:15,2015-01-02 20:05:04,6,0.00,2,N,230,230,4,-52.0,0.0,-0.5,0.00,-8.00,0.3,-60.80
1240537,2,2015-01-04 10:49:09,2015-01-04 10:49:21,5,0.00,2,N,264,50,3,-52.0,0.0,-0.5,0.00,-5.33,0.3,-58.13
1457845,2,2015-01-04 23:46:11,2015-01-04 23:46:25,1,0.00,5,N,265,265,2,-100.0,0.0,-0.5,0.00,-11.75,0.3,-112.55
1457848,2,2015-01-04 23:48:15,2015-01-04 23:48:25,1,0.00,5,N,265,264,2,-100.0,0.0,-0.5,0.00,-11.75,0.3,-112.55
1481154,2,2015-01-05 04:55:14,2015-01-05 04:58:42,1,0.00,5,N,39,39,3,-90.0,0.0,-0.5,0.00,-10.66,0.3,-101.46
2043590,2,2015-01-06 16:21:26,2015-01-06 16:21:38,1,0.00,5,N,1,1,3,-79.0,0.0,-0.5,-12.00,-16.00,0.3,-107.80
2528196,2,2015-01-07 19:42:55,2015-01-07 19:43:16,2,0.00,2,N,132,132,3,-52.0,0.0,-0.5,0.00,-5.33,0.3,-58.13
4116192,2,2015-01-11 02:51:25,2015-01-11 02:51:46,2,0.14,5,N,68,68,4,-150.0,0.0,0.0,0.00,-26.00,0.3,-176.30
4116194,2,2015-01-11 02:53:28,2015-01-11 03:00:42,2,0.00,5,N,68,246,4,-150.0,0.0,0.0,0.00,-26.00,0.3,-176.30
4561572,2,2015-01-12 09:10:40,2015-01-12 09:10:51,1,0.00,2,N,234,234,4,-52.0,0.0,-0.5,0.00,-5.33,0.3,-58.13


In [60]:
df["tolls_amount"].mean()

np.float64(0.24340315832452059)

In [61]:
# df_tolls_amount = df[df["tolls_amount"] < 1]
# quantatity = len(df_tolls_amount)
# percantage = 100 * (quantatity / len(df))
# print(f"1 den az olan köprü geçiş ücretleri: {percantage}")

In [62]:
df["tolls_amount"].max()

np.float64(1450.09)

In [63]:
# plt.figure(figsize=(12,6))
# sns.ecdfplot(data=df, x="tolls_amount", linewidth=3, color="darkred")    # Empirical Cumulative Distribution Function
# plt.axvline(x=0.5, color="blue", linestyle="--", linewidth=2, label="The limit is 0.5") # limit 0.5
# plt.title("the distribution of the tolls amount using a limit")
# plt.xlabel("The price", fontsize=14)
# plt.ylabel("The percantage of the data", fontsize=14)
# plt.legend()
# plt.show()

In [64]:
df["tolls_amount"].quantile(0.99)

np.float64(5.33)

In [65]:
df[df["tolls_amount"] > 1.0]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
17,1,2015-01-01 00:35:34,2015-01-01 00:58:57,1,4.40,1,N,158,145,1,18.2,0.5,0.5,3.00,5.33,0.0,27.83
239,2,2015-01-01 00:58:11,2015-01-01 01:27:27,1,11.88,1,N,163,200,2,36.5,0.5,0.5,0.00,2.70,0.3,40.50
272,2,2015-01-01 00:36:57,2015-01-01 00:54:16,1,6.95,1,N,79,157,2,21.0,0.5,0.5,0.00,5.33,0.3,27.63
334,2,2015-01-01 00:52:37,2015-01-01 01:16:59,1,5.17,1,N,90,112,1,20.0,0.5,0.5,7.75,5.33,0.3,34.38
349,1,2015-01-01 00:25:32,2015-01-01 00:47:29,2,10.90,1,N,236,40,2,31.5,0.5,0.5,0.00,5.33,0.0,38.13
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12740924,1,2015-01-31 23:05:21,2015-01-31 23:30:27,1,9.70,1,N,138,238,1,30.0,0.5,0.5,5.80,5.33,0.3,42.43
12740942,1,2015-01-31 23:31:16,2015-02-01 00:24:51,1,15.20,1,N,48,241,2,48.5,0.5,0.5,0.00,2.44,0.3,52.24
12741009,1,2015-01-31 23:19:19,2015-02-01 00:15:08,3,13.00,1,N,163,220,1,49.5,0.5,0.5,15.00,4.88,0.3,70.68
12741015,2,2015-01-31 23:27:10,2015-02-01 00:07:33,1,12.72,1,N,114,55,1,40.5,0.5,0.5,5.00,5.33,0.3,52.13


In [66]:
df[df["tolls_amount"] > 5.33]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
585,2,2015-01-01 00:20:14,2015-01-01 00:32:02,2,4.38,5,N,246,265,1,45.00,0.0,0.0,10.95,9.75,0.3,66.00
1966,1,2015-01-01 00:15:50,2015-01-01 00:29:01,2,10.40,5,N,262,265,1,65.00,0.0,0.0,0.00,9.75,0.0,75.05
3217,1,2015-01-01 00:51:23,2015-01-01 01:23:15,1,17.30,1,N,25,156,1,49.00,0.5,0.5,0.00,10.66,0.0,60.96
3306,1,2015-01-01 00:27:37,2015-01-01 00:54:26,2,12.00,1,N,246,95,1,35.50,0.5,0.5,8.85,7.50,0.0,53.15
3630,1,2015-01-01 00:27:39,2015-01-01 00:39:54,1,3.90,5,N,50,265,2,50.25,0.0,0.0,0.00,9.75,0.0,60.30
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12736184,2,2015-01-31 23:15:08,2015-01-31 23:36:16,1,9.01,1,N,138,236,2,28.00,0.5,0.5,0.00,10.66,0.3,39.96
12738059,2,2015-01-31 23:11:20,2015-01-31 23:37:09,1,13.23,5,N,116,265,2,70.00,0.0,0.5,0.00,9.75,0.3,80.55
12738569,1,2015-01-31 23:31:31,2015-01-31 23:32:10,5,0.00,5,N,265,265,1,50.00,0.0,0.0,12.00,9.75,0.3,72.05
12739484,1,2015-01-31 23:02:58,2015-01-31 23:39:41,1,29.00,1,N,132,156,2,75.50,0.5,0.5,0.00,10.66,0.3,87.46


In [67]:
# df_tolls_amount = df[df["tolls_amount"] > 5.5]
# quantatity = len(df_tolls_amount)
# percantage = 100 * (quantatity / len(df))
# print(f"5.5 den fazla olan köprü geçiş ücretleri: {percantage}")

In [68]:
df["tolls_amount"].quantile(0.999)

np.float64(11.75)

In [69]:
# df_tolls_amount = df[(df["tolls_amount"] >= 5.5) & (df["tolls_amount"] <= 11.75)]
# quantatity = len(df_tolls_amount)
# percantage = 100 * (quantatity / len(df))
# print(f"5.5 <= x <= 11.75 olan köprü geçiş ücretleri: {percantage}")

In [70]:
df[(df["tolls_amount"] > 5.5) & df["tolls_amount"] < 11.75]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
0,1,2015-01-01 00:11:33,2015-01-01 00:16:48,1,1.0,1,N,41,166,1,5.7,0.5,0.5,1.40,0.0,0.0,8.40
1,1,2015-01-01 00:18:24,2015-01-01 00:24:20,1,0.9,1,N,166,238,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30
2,1,2015-01-01 00:26:19,2015-01-01 00:41:06,1,3.5,1,N,238,162,1,13.2,0.5,0.5,2.90,0.0,0.0,17.40
3,1,2015-01-01 00:45:26,2015-01-01 00:53:20,1,2.1,1,N,162,263,1,8.2,0.5,0.5,2.37,0.0,0.0,11.87
4,1,2015-01-01 00:59:21,2015-01-01 01:05:24,1,1.0,1,N,236,141,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12741030,1,2015-01-31 23:21:42,2015-01-31 23:31:00,1,1.6,1,N,90,249,1,8.0,0.5,0.5,2.32,0.0,0.3,11.62
12741031,1,2015-01-31 23:42:43,2015-01-31 23:49:32,1,0.6,1,N,90,68,1,6.0,0.5,0.5,1.46,0.0,0.3,8.76
12741032,1,2015-01-31 23:55:16,2015-02-01 00:16:45,1,3.0,1,N,68,148,1,15.0,0.5,0.5,4.07,0.0,0.3,20.37
12741033,1,2015-01-31 23:20:53,2015-02-01 00:07:35,1,6.9,1,N,189,237,1,32.5,0.5,0.5,6.00,0.0,0.3,39.80


In [71]:
df[df["tolls_amount"] > 12]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
4050,1,2015-01-01 00:13:21,2015-01-01 00:41:33,1,13.90,1,N,87,206,2,41.0,0.5,0.5,0.00,15.99,0.0,58.29
14394,1,2015-01-01 00:43:22,2015-01-01 01:22:31,2,18.10,1,N,158,176,1,52.0,0.5,0.5,0.00,15.99,0.0,69.29
24032,2,2015-01-01 00:37:44,2015-01-01 01:21:22,1,19.41,1,N,162,172,2,55.0,0.5,0.5,0.00,15.99,0.3,72.29
27508,1,2015-01-01 00:22:52,2015-01-01 01:11:32,1,16.50,1,N,252,265,2,53.0,0.5,0.5,0.00,15.08,0.0,69.38
30425,1,2015-01-01 01:41:05,2015-01-01 02:22:47,1,18.50,3,N,112,1,1,72.0,0.5,0.0,0.00,14.75,0.0,87.55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12691070,1,2015-01-31 22:08:22,2015-01-31 22:14:50,1,1.60,1,N,82,260,2,7.5,0.5,0.5,0.00,29.00,0.3,37.80
12712293,2,2015-01-31 23:35:45,2015-02-01 00:07:11,1,20.10,5,N,10,265,2,100.0,0.0,0.5,0.00,15.08,0.3,115.88
12713338,2,2015-01-31 23:12:05,2015-01-31 23:48:11,1,23.49,1,N,132,265,1,62.0,0.5,0.5,23.87,17.08,0.3,104.25
12717181,1,2015-01-31 23:08:26,2015-02-01 00:00:53,1,27.50,1,N,230,204,2,76.0,0.5,0.5,0.00,15.99,0.3,93.29


In [72]:
# df_tolls_amount = df[df["tolls_amount"] > 12]
# quantatity = len(df_tolls_amount)
# percantage = 100 * (quantatity / len(df))
# print(f"12 den fazla olan köprü geçiş ücretleri: {percantage}")

In [73]:
# df_tolls_amount = df[df["tolls_amount"] < 12]
# quantatity = len(df_tolls_amount)
# percantage = 100 * (quantatity / len(df))
# print(f"12 den az olan köprü geçiş ücretlerinin yüzdesi: {percantage}")

In [74]:
df["tolls_amount"].max()

np.float64(1450.09)

Sonuç: 1450$ gibi uçuk rakamlarla çalışmaya gerek yok, aralık 0 < x < 12 olabilir.

### 2.Duration

In [75]:
df.dtypes

VendorID                          int64
tpep_pickup_datetime     datetime64[us]
tpep_dropoff_datetime    datetime64[us]
passenger_count                   int64
trip_distance                   float64
RatecodeID                        int64
store_and_fwd_flag               object
PULocationID                      int64
DOLocationID                      int64
payment_type                      int64
fare_amount                     float64
extra                           float64
mta_tax                         float64
tip_amount                      float64
tolls_amount                    float64
improvement_surcharge           float64
total_amount                    float64
dtype: object

In [76]:
df["duration"] = (df["tpep_dropoff_datetime"] - df["tpep_pickup_datetime"]).dt.total_seconds() / 60
# dakika bazında geçen süreyi yeni sütun olarak ekledim
df

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,duration
0,1,2015-01-01 00:11:33,2015-01-01 00:16:48,1,1.0,1,N,41,166,1,5.7,0.5,0.5,1.40,0.0,0.0,8.40,5.250000
1,1,2015-01-01 00:18:24,2015-01-01 00:24:20,1,0.9,1,N,166,238,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30,5.933333
2,1,2015-01-01 00:26:19,2015-01-01 00:41:06,1,3.5,1,N,238,162,1,13.2,0.5,0.5,2.90,0.0,0.0,17.40,14.783333
3,1,2015-01-01 00:45:26,2015-01-01 00:53:20,1,2.1,1,N,162,263,1,8.2,0.5,0.5,2.37,0.0,0.0,11.87,7.900000
4,1,2015-01-01 00:59:21,2015-01-01 01:05:24,1,1.0,1,N,236,141,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30,6.050000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12741030,1,2015-01-31 23:21:42,2015-01-31 23:31:00,1,1.6,1,N,90,249,1,8.0,0.5,0.5,2.32,0.0,0.3,11.62,9.300000
12741031,1,2015-01-31 23:42:43,2015-01-31 23:49:32,1,0.6,1,N,90,68,1,6.0,0.5,0.5,1.46,0.0,0.3,8.76,6.816667
12741032,1,2015-01-31 23:55:16,2015-02-01 00:16:45,1,3.0,1,N,68,148,1,15.0,0.5,0.5,4.07,0.0,0.3,20.37,21.483333
12741033,1,2015-01-31 23:20:53,2015-02-01 00:07:35,1,6.9,1,N,189,237,1,32.5,0.5,0.5,6.00,0.0,0.3,39.80,46.700000


In [77]:
df[df["duration"] <= 0]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,duration
218,1,2015-01-01 00:39:43,2015-01-01 00:39:43,2,0.0,1,N,113,264,2,14.0,0.5,0.5,0.0,0.0,0.0,15.3,0.0
364,1,2015-01-01 00:52:34,2015-01-01 00:52:34,1,0.0,1,N,264,264,2,4.5,0.5,0.5,0.0,0.0,0.0,5.8,0.0
655,1,2015-01-01 00:39:41,2015-01-01 00:39:41,1,0.0,1,N,236,264,2,2.5,0.5,0.5,0.0,0.0,0.0,3.8,0.0
2981,1,2015-01-01 00:34:00,2015-01-01 00:34:00,2,0.0,1,N,90,264,2,21.0,0.5,0.5,0.0,0.0,0.0,22.3,0.0
6538,1,2015-01-01 00:17:14,2015-01-01 00:17:14,1,0.0,1,N,238,264,2,10.5,0.5,0.5,0.0,0.0,0.0,11.8,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12736787,1,2015-01-31 23:08:56,2015-01-31 23:08:56,1,0.0,1,N,113,264,2,17.5,0.5,0.5,0.0,0.0,0.3,18.8,0.0
12737785,1,2015-01-31 23:36:58,2015-01-31 23:36:58,1,0.0,1,N,164,264,2,2.5,0.5,0.5,0.0,0.0,0.3,3.8,0.0
12738073,1,2015-01-31 23:54:14,2015-01-31 23:54:14,1,0.0,1,N,114,264,2,2.5,0.5,0.5,0.0,0.0,0.3,3.8,0.0
12738074,1,2015-01-31 23:54:16,2015-01-31 23:54:16,1,0.0,1,N,114,264,2,2.5,0.5,0.5,0.0,0.0,0.3,3.8,0.0


İzlenim: yolcuların binme-inme datetime değerlerini kullanarak arada geçen süre ile yeni bir sütun oluşturdum.

In [78]:
# df_nonpositive_duration = df[df["duration"] <= 0]
# quantatity_duration = len(df_nonpositive_duration)
# percantage = 100 * (quantatity_duration / len(df))
# print(f"Yolculuk süresi sifirdan az veya eşit olanlarin yüzdesi: {percantage}")

In [79]:
df["duration"].min()

np.float64(-1211.0166666666667)

In [80]:
df["duration"].max()

np.float64(548555.6333333333)

İzlenim: Bir insanın 548555 dk taksi yapması imkansıza yakın bir şey idir.

In [81]:
df["duration"].quantile(0.999)

np.float64(101.48658333336934)

In [82]:
# df_duration = df[df["duration"] <= 102]
# percantage = 100 * (len(df_duration) / len(df))
# print(f"102 dk veya daha az taksi yapma yüzdesi: {percantage}")

In [83]:
df[df["duration"] > 102]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,duration
2335,2,2015-01-01 00:37:12,2015-01-02 00:28:04,1,4.53,1,N,48,144,1,21.5,0.5,0.5,0.00,0.0,0.3,22.80,1430.866667
3160,1,2015-01-01 00:18:54,2015-01-01 02:30:04,1,2.60,1,Y,48,188,2,11.0,0.0,0.5,0.00,0.0,0.0,11.80,131.166667
4350,2,2015-01-01 00:08:28,2015-01-02 00:00:00,1,2.78,1,N,231,186,1,11.0,0.5,0.5,0.00,0.0,0.3,12.30,1431.533333
4813,2,2015-01-01 00:35:44,2015-01-02 00:34:09,2,1.44,1,N,249,137,2,7.5,0.5,0.5,0.00,0.0,0.3,8.80,1438.416667
5558,1,2015-01-01 00:05:31,2015-01-01 03:47:46,1,32.90,1,N,48,265,2,2.5,0.5,0.5,0.00,0.0,0.0,3.80,222.250000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12734574,1,2015-01-31 23:13:22,2015-02-01 05:33:44,1,18.20,1,N,132,97,1,49.0,0.5,0.5,10.06,0.0,0.3,60.36,380.366667
12735965,2,2015-01-31 23:41:19,2015-02-01 23:40:50,2,3.05,1,N,230,262,1,12.5,0.5,0.5,1.30,0.0,0.3,15.10,1439.516667
12736500,2,2015-01-31 23:47:10,2015-02-01 23:21:05,1,10.48,1,N,79,244,1,40.5,0.5,0.5,8.36,0.0,0.3,50.16,1413.916667
12737677,2,2015-01-31 23:02:45,2015-02-01 22:53:00,1,7.63,1,N,97,161,1,27.0,0.5,0.5,1.42,0.0,0.3,29.72,1430.250000


Sonuç: 0 < x <= 102 "duration" için makul sınırlardır.

In [84]:
df.head()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,duration
0,1,2015-01-01 00:11:33,2015-01-01 00:16:48,1,1.0,1,N,41,166,1,5.7,0.5,0.5,1.40,0.0,0.0,8.40,5.250000
1,1,2015-01-01 00:18:24,2015-01-01 00:24:20,1,0.9,1,N,166,238,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30,5.933333
2,1,2015-01-01 00:26:19,2015-01-01 00:41:06,1,3.5,1,N,238,162,1,13.2,0.5,0.5,2.90,0.0,0.0,17.40,14.783333
3,1,2015-01-01 00:45:26,2015-01-01 00:53:20,1,2.1,1,N,162,263,1,8.2,0.5,0.5,2.37,0.0,0.0,11.87,7.900000
4,1,2015-01-01 00:59:21,2015-01-01 01:05:24,1,1.0,1,N,236,141,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30,6.050000


### 3. Passanger Count

In [85]:
df[df["passenger_count"] <= 0]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,duration
649,1,2015-01-01 00:29:56,2015-01-01 00:38:38,0,1.3,1,N,164,79,1,7.50,0.5,0.5,2.00,0.0,0.0,10.80,8.700000
651,1,2015-01-01 00:58:49,2015-01-01 01:14:50,0,2.8,1,N,186,261,2,13.00,0.5,0.5,0.00,0.0,0.0,14.30,16.016667
16259,1,2015-01-01 00:42:58,2015-01-01 01:05:35,0,4.3,1,N,186,261,2,18.00,0.5,0.5,0.00,0.0,0.0,19.30,22.616667
21690,1,2015-01-01 00:38:54,2015-01-01 04:38:54,0,0.0,99,Y,79,264,1,10.55,0.0,0.0,0.00,0.0,0.0,10.55,240.000000
22085,2,2015-01-01 00:56:57,2015-01-01 00:57:10,0,0.0,5,N,145,145,1,26.33,0.0,0.5,0.00,0.0,0.3,27.13,0.216667
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12736224,1,2015-01-31 23:05:44,2015-01-31 23:21:12,0,4.4,1,N,163,166,2,15.00,0.5,0.5,0.00,0.0,0.3,16.30,15.466667
12736225,1,2015-01-31 23:30:28,2015-01-31 23:39:26,0,1.9,1,N,238,41,1,9.00,0.5,0.5,2.05,0.0,0.3,12.35,8.966667
12738216,1,2015-01-31 23:49:11,2015-01-31 23:53:37,0,0.1,5,N,132,132,4,0.01,0.0,0.0,0.00,0.0,0.3,0.31,4.433333
12739471,1,2015-01-31 23:10:15,2015-01-31 23:24:03,0,1.8,1,N,246,79,1,10.00,0.5,0.5,2.00,0.0,0.3,13.30,13.800000


In [86]:
df["passenger_count"].max()

np.int64(9)

In [87]:
df["passenger_count"].quantile(0.999)

np.float64(6.0)

In [88]:
df[(df["passenger_count"] > 0) & (df["passenger_count"] <= 6)]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,duration
0,1,2015-01-01 00:11:33,2015-01-01 00:16:48,1,1.0,1,N,41,166,1,5.7,0.5,0.5,1.40,0.0,0.0,8.40,5.250000
1,1,2015-01-01 00:18:24,2015-01-01 00:24:20,1,0.9,1,N,166,238,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30,5.933333
2,1,2015-01-01 00:26:19,2015-01-01 00:41:06,1,3.5,1,N,238,162,1,13.2,0.5,0.5,2.90,0.0,0.0,17.40,14.783333
3,1,2015-01-01 00:45:26,2015-01-01 00:53:20,1,2.1,1,N,162,263,1,8.2,0.5,0.5,2.37,0.0,0.0,11.87,7.900000
4,1,2015-01-01 00:59:21,2015-01-01 01:05:24,1,1.0,1,N,236,141,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30,6.050000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12741030,1,2015-01-31 23:21:42,2015-01-31 23:31:00,1,1.6,1,N,90,249,1,8.0,0.5,0.5,2.32,0.0,0.3,11.62,9.300000
12741031,1,2015-01-31 23:42:43,2015-01-31 23:49:32,1,0.6,1,N,90,68,1,6.0,0.5,0.5,1.46,0.0,0.3,8.76,6.816667
12741032,1,2015-01-31 23:55:16,2015-02-01 00:16:45,1,3.0,1,N,68,148,1,15.0,0.5,0.5,4.07,0.0,0.3,20.37,21.483333
12741033,1,2015-01-31 23:20:53,2015-02-01 00:07:35,1,6.9,1,N,189,237,1,32.5,0.5,0.5,6.00,0.0,0.3,39.80,46.700000


In [89]:
percantage = (12734227 / len(df)) * 100
print(f"0 ve 6 arasindaki yolcu sayilarinin yüzdesi: {percantage}")

0 ve 6 arasindaki yolcu sayilarinin yüzdesi: 99.94836276449072


Sonuç: yolcu sayıları 0 ve 6 arasında (6 dahil) olanlar baz alınacaktır.

### 4. Trip Distance

In [90]:
df["trip_distance"].min()

np.float64(0.0)

In [91]:
df["trip_distance"].max()

np.float64(15420004.5)

In [92]:
df["trip_distance"].quantile(0.999)

np.float64(23.1)

In [93]:
df[df["trip_distance"] > 24]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,duration
714,1,2015-01-01 00:17:51,2015-01-01 00:58:34,1,25.80,1,N,43,29,1,68.7,0.5,0.5,14.00,0.00,0.0,84.00,40.716667
4777,2,2015-01-01 00:34:31,2015-01-01 01:33:33,5,26.93,1,N,132,93,2,75.0,0.5,0.5,0.00,0.00,0.3,76.30,59.033333
4809,2,2015-01-01 00:38:36,2015-01-01 01:16:59,1,24.94,1,N,132,200,1,65.5,0.5,0.5,5.00,7.77,0.3,79.57,38.383333
5558,1,2015-01-01 00:05:31,2015-01-01 03:47:46,1,32.90,1,N,48,265,2,2.5,0.5,0.5,0.00,0.00,0.0,3.80,222.250000
10830,1,2015-01-01 00:33:41,2015-01-01 01:12:22,2,24.50,1,N,237,55,1,65.2,0.5,0.5,3.50,0.00,0.0,70.00,38.683333
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12738807,1,2015-01-31 23:40:00,2015-02-01 00:34:56,1,28.80,1,N,140,265,1,80.0,0.5,0.5,16.25,0.00,0.3,97.55,54.933333
12739292,1,2015-01-31 23:55:19,2015-02-01 00:51:27,1,29.10,5,N,100,265,1,125.0,0.0,0.0,31.32,0.00,0.3,156.62,56.133333
12739484,1,2015-01-31 23:02:58,2015-01-31 23:39:41,1,29.00,1,N,132,156,2,75.5,0.5,0.5,0.00,10.66,0.3,87.46,36.716667
12739695,2,2015-01-31 23:24:34,2015-02-01 00:00:46,2,24.71,1,N,132,228,2,65.0,0.5,0.5,0.00,0.00,0.3,66.30,36.200000


In [94]:
df[(df["trip_distance"] > 0) & (df["trip_distance"] <= 24)]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,duration
0,1,2015-01-01 00:11:33,2015-01-01 00:16:48,1,1.0,1,N,41,166,1,5.7,0.5,0.5,1.40,0.0,0.0,8.40,5.250000
1,1,2015-01-01 00:18:24,2015-01-01 00:24:20,1,0.9,1,N,166,238,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30,5.933333
2,1,2015-01-01 00:26:19,2015-01-01 00:41:06,1,3.5,1,N,238,162,1,13.2,0.5,0.5,2.90,0.0,0.0,17.40,14.783333
3,1,2015-01-01 00:45:26,2015-01-01 00:53:20,1,2.1,1,N,162,263,1,8.2,0.5,0.5,2.37,0.0,0.0,11.87,7.900000
4,1,2015-01-01 00:59:21,2015-01-01 01:05:24,1,1.0,1,N,236,141,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30,6.050000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12741030,1,2015-01-31 23:21:42,2015-01-31 23:31:00,1,1.6,1,N,90,249,1,8.0,0.5,0.5,2.32,0.0,0.3,11.62,9.300000
12741031,1,2015-01-31 23:42:43,2015-01-31 23:49:32,1,0.6,1,N,90,68,1,6.0,0.5,0.5,1.46,0.0,0.3,8.76,6.816667
12741032,1,2015-01-31 23:55:16,2015-02-01 00:16:45,1,3.0,1,N,68,148,1,15.0,0.5,0.5,4.07,0.0,0.3,20.37,21.483333
12741033,1,2015-01-31 23:20:53,2015-02-01 00:07:35,1,6.9,1,N,189,237,1,32.5,0.5,0.5,6.00,0.0,0.3,39.80,46.700000


In [95]:
df[df["trip_distance"] > 24]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,duration
714,1,2015-01-01 00:17:51,2015-01-01 00:58:34,1,25.80,1,N,43,29,1,68.7,0.5,0.5,14.00,0.00,0.0,84.00,40.716667
4777,2,2015-01-01 00:34:31,2015-01-01 01:33:33,5,26.93,1,N,132,93,2,75.0,0.5,0.5,0.00,0.00,0.3,76.30,59.033333
4809,2,2015-01-01 00:38:36,2015-01-01 01:16:59,1,24.94,1,N,132,200,1,65.5,0.5,0.5,5.00,7.77,0.3,79.57,38.383333
5558,1,2015-01-01 00:05:31,2015-01-01 03:47:46,1,32.90,1,N,48,265,2,2.5,0.5,0.5,0.00,0.00,0.0,3.80,222.250000
10830,1,2015-01-01 00:33:41,2015-01-01 01:12:22,2,24.50,1,N,237,55,1,65.2,0.5,0.5,3.50,0.00,0.0,70.00,38.683333
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12738807,1,2015-01-31 23:40:00,2015-02-01 00:34:56,1,28.80,1,N,140,265,1,80.0,0.5,0.5,16.25,0.00,0.3,97.55,54.933333
12739292,1,2015-01-31 23:55:19,2015-02-01 00:51:27,1,29.10,5,N,100,265,1,125.0,0.0,0.0,31.32,0.00,0.3,156.62,56.133333
12739484,1,2015-01-31 23:02:58,2015-01-31 23:39:41,1,29.00,1,N,132,156,2,75.5,0.5,0.5,0.00,10.66,0.3,87.46,36.716667
12739695,2,2015-01-31 23:24:34,2015-02-01 00:00:46,2,24.71,1,N,132,228,2,65.0,0.5,0.5,0.00,0.00,0.3,66.30,36.200000


In [96]:
100 * (12657215 / len(df))  # trip_distance 24 ve ALTINDA olanların yüzdesi

99.34391120938503

Sonuç: 0 < x <= 24  trip_distance için makul aralıklar

### 5. RateCode ID

In [98]:
df[df["RatecodeID"] > 6]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,duration
21690,1,2015-01-01 00:38:54,2015-01-01 04:38:54,0,0.0,99,Y,79,264,1,10.55,0.0,0.0,0.00,0.0,0.0,10.55,240.000000
39003,1,2015-01-01 01:16:26,2015-01-01 02:30:00,1,7.5,99,Y,239,237,1,29.50,0.5,0.5,0.00,0.0,0.0,30.50,73.566667
45586,1,2015-01-01 01:49:52,2015-01-01 01:49:55,1,0.0,99,N,61,61,2,0.00,0.0,0.0,0.00,0.0,0.0,0.30,0.050000
55934,1,2015-01-01 01:57:00,2015-01-01 02:00:31,1,0.6,99,Y,236,262,1,4.50,0.5,0.5,1.00,0.0,0.0,6.50,3.516667
125278,1,2015-01-01 04:53:45,2015-01-01 08:53:45,0,0.0,99,Y,112,264,3,0.00,0.0,0.0,0.00,0.0,0.0,0.00,240.000000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12546750,1,2015-01-31 17:14:31,2015-01-31 17:27:37,1,2.5,99,Y,144,186,1,11.50,0.0,0.5,2.00,0.0,0.3,14.30,13.100000
12569342,1,2015-01-31 18:10:37,2015-01-31 22:10:37,0,0.0,99,Y,48,264,1,14.30,0.0,0.0,0.00,0.0,0.0,14.30,240.000000
12631665,1,2015-01-31 20:48:18,2015-01-31 20:55:50,1,0.8,99,Y,90,249,1,6.00,0.5,0.5,1.82,0.0,0.3,9.12,7.533333
12636083,1,2015-01-31 20:17:52,2015-02-01 00:17:52,0,0.0,99,Y,79,264,1,6.80,0.0,0.0,0.00,0.0,0.0,6.80,240.000000


In [99]:
df[df["RatecodeID"] == 99]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,duration
21690,1,2015-01-01 00:38:54,2015-01-01 04:38:54,0,0.0,99,Y,79,264,1,10.55,0.0,0.0,0.00,0.0,0.0,10.55,240.000000
39003,1,2015-01-01 01:16:26,2015-01-01 02:30:00,1,7.5,99,Y,239,237,1,29.50,0.5,0.5,0.00,0.0,0.0,30.50,73.566667
45586,1,2015-01-01 01:49:52,2015-01-01 01:49:55,1,0.0,99,N,61,61,2,0.00,0.0,0.0,0.00,0.0,0.0,0.30,0.050000
55934,1,2015-01-01 01:57:00,2015-01-01 02:00:31,1,0.6,99,Y,236,262,1,4.50,0.5,0.5,1.00,0.0,0.0,6.50,3.516667
125278,1,2015-01-01 04:53:45,2015-01-01 08:53:45,0,0.0,99,Y,112,264,3,0.00,0.0,0.0,0.00,0.0,0.0,0.00,240.000000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12546750,1,2015-01-31 17:14:31,2015-01-31 17:27:37,1,2.5,99,Y,144,186,1,11.50,0.0,0.5,2.00,0.0,0.3,14.30,13.100000
12569342,1,2015-01-31 18:10:37,2015-01-31 22:10:37,0,0.0,99,Y,48,264,1,14.30,0.0,0.0,0.00,0.0,0.0,14.30,240.000000
12631665,1,2015-01-31 20:48:18,2015-01-31 20:55:50,1,0.8,99,Y,90,249,1,6.00,0.5,0.5,1.82,0.0,0.3,9.12,7.533333
12636083,1,2015-01-31 20:17:52,2015-02-01 00:17:52,0,0.0,99,Y,79,264,1,6.80,0.0,0.0,0.00,0.0,0.0,6.80,240.000000


İzlenim: Klasik RateCodeID ler dışında geri kalan eskik ya da diğer diyebileceğimiz veriler "99" ile kodlanmış

In [100]:
100 * (498 / len(df))   # 99 olanların yüzdesi

0.003908700909502899

In [101]:
df

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,duration
0,1,2015-01-01 00:11:33,2015-01-01 00:16:48,1,1.0,1,N,41,166,1,5.7,0.5,0.5,1.40,0.0,0.0,8.40,5.250000
1,1,2015-01-01 00:18:24,2015-01-01 00:24:20,1,0.9,1,N,166,238,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30,5.933333
2,1,2015-01-01 00:26:19,2015-01-01 00:41:06,1,3.5,1,N,238,162,1,13.2,0.5,0.5,2.90,0.0,0.0,17.40,14.783333
3,1,2015-01-01 00:45:26,2015-01-01 00:53:20,1,2.1,1,N,162,263,1,8.2,0.5,0.5,2.37,0.0,0.0,11.87,7.900000
4,1,2015-01-01 00:59:21,2015-01-01 01:05:24,1,1.0,1,N,236,141,3,6.0,0.5,0.5,0.00,0.0,0.0,7.30,6.050000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12741030,1,2015-01-31 23:21:42,2015-01-31 23:31:00,1,1.6,1,N,90,249,1,8.0,0.5,0.5,2.32,0.0,0.3,11.62,9.300000
12741031,1,2015-01-31 23:42:43,2015-01-31 23:49:32,1,0.6,1,N,90,68,1,6.0,0.5,0.5,1.46,0.0,0.3,8.76,6.816667
12741032,1,2015-01-31 23:55:16,2015-02-01 00:16:45,1,3.0,1,N,68,148,1,15.0,0.5,0.5,4.07,0.0,0.3,20.37,21.483333
12741033,1,2015-01-31 23:20:53,2015-02-01 00:07:35,1,6.9,1,N,189,237,1,32.5,0.5,0.5,6.00,0.0,0.3,39.80,46.700000


In [102]:
df.dtypes

VendorID                          int64
tpep_pickup_datetime     datetime64[us]
tpep_dropoff_datetime    datetime64[us]
passenger_count                   int64
trip_distance                   float64
RatecodeID                        int64
store_and_fwd_flag               object
PULocationID                      int64
DOLocationID                      int64
payment_type                      int64
fare_amount                     float64
extra                           float64
mta_tax                         float64
tip_amount                      float64
tolls_amount                    float64
improvement_surcharge           float64
total_amount                    float64
duration                        float64
dtype: object

In [104]:
df[df["store_and_fwd_flag"] == "Y"]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,duration
31,1,2015-01-01 00:19:23,2015-01-01 00:38:24,1,7.7,1,Y,231,236,1,23.7,0.5,0.5,5.00,0.0,0.0,30.00,19.016667
58,1,2015-01-01 00:30:11,2015-01-01 00:37:02,2,1.4,1,Y,244,243,2,7.0,0.5,0.5,0.00,0.0,0.0,8.30,6.850000
72,1,2015-01-01 00:12:05,2015-01-01 00:16:44,1,0.7,1,Y,164,107,1,5.0,0.5,0.5,1.25,0.0,0.0,7.55,4.650000
310,1,2015-01-01 00:17:30,2015-01-01 00:25:33,1,1.6,1,Y,4,144,2,8.0,0.5,0.5,0.00,0.0,0.0,9.30,8.050000
402,1,2015-01-01 00:47:33,2015-01-01 01:06:40,1,2.6,1,Y,246,143,2,14.0,0.5,0.5,0.00,0.0,0.0,15.30,19.116667
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12740895,1,2015-01-31 23:27:32,2015-01-31 23:40:51,1,2.7,1,Y,236,50,1,11.5,0.5,0.5,2.56,0.0,0.3,15.36,13.316667
12740943,1,2015-01-31 23:22:47,2015-01-31 23:32:30,1,1.9,1,Y,74,151,2,9.0,0.5,0.5,0.00,0.0,0.3,10.30,9.716667
12740944,1,2015-01-31 23:42:08,2015-02-01 00:04:01,1,4.3,1,Y,264,48,2,17.5,0.5,0.5,0.00,0.0,0.3,18.80,21.883333
12740985,1,2015-01-31 23:42:54,2015-01-31 23:59:52,1,3.8,1,Y,114,33,2,15.5,0.5,0.5,0.00,0.0,0.3,16.80,16.966667


İzlenim: "store_and_fwd_flag" verisi veri akışında kesinti olup olmadığını söyler. Veri yapısını bozmadığından ignore edilebilir.

In [106]:
df[df["PULocationID"] == df["DOLocationID"]]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,duration
20,1,2015-01-01 00:23:14,2015-01-01 00:34:15,1,0.40,1,N,50,50,2,7.5,0.5,0.5,0.00,0.0,0.0,8.80,11.016667
30,1,2015-01-01 00:11:14,2015-01-01 00:11:34,1,0.00,1,N,114,114,2,2.5,0.5,0.5,0.00,0.0,0.0,3.80,0.333333
60,1,2015-01-01 00:56:38,2015-01-01 00:58:53,2,0.40,1,N,244,244,2,3.5,0.5,0.5,0.00,0.0,0.0,4.80,2.250000
62,1,2015-01-01 00:18:01,2015-01-01 00:37:22,1,3.00,1,N,264,264,2,14.5,0.5,0.5,0.00,0.0,0.0,15.80,19.350000
63,1,2015-01-01 00:46:38,2015-01-01 00:48:28,1,0.20,1,N,264,264,2,3.0,0.5,0.5,0.00,0.0,0.0,4.30,1.833333
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12740966,2,2015-01-31 23:10:01,2015-01-31 23:10:32,1,0.05,2,N,264,264,2,52.0,0.0,0.5,0.00,0.0,0.3,52.80,0.516667
12740968,2,2015-01-31 23:29:06,2015-01-31 23:29:08,1,0.01,2,N,264,264,2,52.0,0.0,0.5,0.00,0.0,0.3,52.80,0.033333
12740973,1,2015-01-31 23:26:29,2015-01-31 23:28:03,3,0.30,1,N,239,239,2,3.5,0.5,0.5,0.00,0.0,0.3,4.80,1.566667
12740975,1,2015-01-31 23:57:34,2015-02-01 00:02:55,1,0.70,1,N,107,107,1,5.5,0.5,0.5,1.36,0.0,0.3,8.16,5.350000
